### ISSUE TITLE
Building a controller for Browser Extension

### ISSUE THREAD

**Issue by Linzo99** [Dankovk/browser-use-node#1]
This is really cool, what do you thing about putting the package into a browser extension (Ex: chrome extension) and let it control your running browser rather than a playwright instance ?

Just like Project Mariner.

**Comment by Dankovk** [comment]
Hey @Linzo99 ! What a great idea! 
But there are some open questions here: 
- this lib has multi-agent and multi-tab support, how well can we achieve it in the context of browser extension
- Any suggestions for browser automation lib or tool that we can use, that will be able to wrap the main page/browser/network operations under some simple API like playwright's. I was thinking about just using web-view (bun has some cool new implementations)
- We will loose api compatibility with the original python lib? (I'm now trying to preserve as far as I can)
- Security and anti-tracking hacks and patches? The current lib logic packed with a lot of them, and some of them are possible because we can start a browser with chromium flags to allow or disable somerthing


But overall I'm a huge fan of a such approach

**Comment by Linzo99** [comment]
Yeah it's true, that may involve loosing some compatibility with the python lib.

I was thinking about using the [chromeExtensionAsync](https://github.com/KeithHenry/chromeExtensionAsync) to mimic some of playwright methods and JS to implement the actions and page related functions.
- But i'm not sure if the "multi-agent" will be supported
- Multi-tab, i think that may be possible..
- About Security, if you're talking about the anti-detection, since it will be running as an extension, i think that will not be a problem
- About anti-tracking maybe block some request pattern, or just let the browser or another extension handle it.

I think this will be the next iteration of web automation agents, having them in your daily browser as an extension.

**Comment by vtempest** [comment]
Spasibo za vse Alex!

We're using it in sidebar on chrome ext at qwksearch. Code in my ai research agent repo

https://pptr.dev/guides/running-puppeteer-in-extensions
You can run ppptr in crx and even python w provide

Just keep up to date with the controlling code and transfer changes

Multitab is easy with chrome api look into my ext for demo

It's possible with bg service task runner and web api calls
